File: multiple_linear_regression/multiple_linear_regression.py
# multiple linear regression by 2blam

# import the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


# import dataset
dataset = pd.read_csv("50_Startups.csv")
X = dataset.iloc[:, :-1].values #R&D Spend, Administration, Marketing Spend, State
y = dataset.iloc[:, 4].values #Profit

# encode categorical data ?
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
encodeColIdx = 3 # 3 == State column
X[:, encodeColIdx] = labelencoder_X.fit_transform(X[:, encodeColIdx]) # State--> 0, 1, 2
onehotencoder = OneHotEncoder(categorical_features=[encodeColIdx]) # create dummy variable representation, i.e. 0 --> 1 0 0; 1 --> 0 1 0; 2 --> 0 0 1
X = onehotencoder.fit_transform(X).toarray()

# avoid the dummy variable trap
X = X[:, 1:] # remove the 1st column (i.e. remove 1 dummy variable)
# NOTE:
#  even we don't not provide the above statement,
#  the library would handle about the dummy variable trap

# split into training and testing dataset
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=0) # 20% for testing dataset

# multiple linear regression
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# predict
y_pred = regressor.predict(X_test)

File: multiple_linear_regression/multiple_linear_regression.R
# multiple linear regression (by 2blam)

setwd("~/Documents/Git/ML/multiple_linear_regression")

# import dataset
dataset = read.csv("50_Startups.csv")

# encode categorical data
dataset$State = factor(dataset$State, 
                         levels=unique(dataset$State),
                         labels=c(1, 2, 3))

# split into training and testing dataset
#install.packages("caTools") - first time installation
library(caTools) # for splitting dataset
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split)
test_set = subset(dataset, !split)

# multiple linear regression
regressor = lm(formula = Profit ~ ., data=training_set) #use . to represent using the rest of the predictors
summary(regressor)

#Note:
# In Coefficients, 
# 1) it always shows (Intercept) and R.D.Spend with "***". 
#    In other words, R.D.Spend is the most powerful predictor for profit
# 2) lm auto remove 1 dummy variable, i.e. avoid dummy variable trap

# predict
y_pred =predict(regressor, newdata=test_set)
